ecosystem/indexer-grpc/indexer-grpc-file-store-data-integrity-checker/src/worker.rs

```diff
@@ -0,0 +1,84 @@
+use tokio::sync::Mutex;
+
+pub(crate) struct Worker {
+    file_store_config: IndexerGrpcFileStoreConfig,
+    starting_version: Option<u64>,
+}
+
+impl Worker {
+    pub async fn new(
+        file_store_config: IndexerGrpcFileStoreConfig,
+        starting_version: Option<u64>,
+    ) -> Result<Self> {
+        Ok(Self {
+            file_store_config,
+            starting_version,
+        })
+    }
+
+    pub async fn run(&mut self) -> Result<()> {
+        let file_store_operator = self.file_store_config.create();
+        let starting_vresion = self.starting_version.unwrap_or(0);
+        let cursor = Arc::new(Mutex::new(starting_vresion));
+        let data_to_process = Arc::new(Mutex::new(BTreeMap::new()));
+        let task_count = 32;
+        let current_metadata = file_store_operator.get_file_store_metadata().await.unwrap();
+        let ending_version = current_metadata.version;
+        for _ in 0..task_count {
+            let file_store_operator = file_store_operator.clone_box();
+            let cursor = cursor.clone();
+            let data_to_process = data_to_process.clone();
+            tokio::spawn(async move {
+                loop {
+                    let version_to_process = {
+                        let mut cursor_now = cursor.lock().await;
+                        let version_to_process = *cursor_now;
+                        *cursor_now += 1000;
+                        version_to_process
+                    };
+                    if version_to_process >= ending_version {
+                        break;
+                    }
+
+                    let transactions = file_store_operator
+                        .get_transactions(version_to_process, 3)
+                        .await
+                        .unwrap();
+                    let mut data_to_process_now = data_to_process.lock().await;
+                    data_to_process_now.insert(version_to_process, transactions);
+                }
+            });
+        }
+        // Checker loop.
+        let mut version_to_check = 0;
+        loop {
+            if version_to_check >= ending_version {
+                break;
+            }
+            // Get the lowest version in data_to_process.
+            let transactions = {
+                let mut data_to_process_now = data_to_process.lock().await;
+                if !data_to_process_now.contains_key(&version_to_check) {
+                    tokio::time::sleep(tokio::time::Duration::from_millis(20)).await;
+                    continue;
+                }
+                data_to_process_now.remove(&version_to_check).unwrap()
+            };
+            for transaction in transactions {
+                assert_eq!(transaction.version, version_to_check);
+                version_to_check += 1;
+            }
+        }
+
+        loop {
+            tokio::time::sleep(tokio::time::Duration::from_secs(60)).await;
+            tracing::info!("Data Integrity Checker is done!");
+        }
```

Comment:
nit: maybe we can add a flag like `is_processing_complete` that could be updated elsewhere in the code to indicate that the processing tasks are done, and possibly use this to exit the loop. 

```
let is_processing_complete = Arc::new(AtomicBool::new(false));

loop {
    tokio::time::sleep(Duration::from_secs(60)).await;
    if is_processing_complete.load(Ordering::Acquire) {
        tracing::info!("Data Integrity Checker has completed processing. Exiting loop.");
        break;
    } else {
        // Example of a more informative log message
        tracing::info!("Data Integrity Checker is running. Waiting for processing to complete.");
    }
}
```

Reply:
yea; but the concern is if you don't sleep after complete, job will start from 0 again 🤦

Reply:
ohhh make sense. ignore this 😛